Create module for providing client-server communication between pokemon.

ijassiem opened the pull request:
This module was created to provide a means for two pokemon objects to communicate with
each other through sockets. This module creates a new Comms class which inherits from
the ElectricPokemon class. Co-routines are added to the class for implenting client-server
communication through sockets and asyncio. A minor change was also made to the pokemon
class in pokemon.py.

Contributes to: DGDT-35

ijassiem committed:
Create module for providing client-server communication between pokemon.

This module was created to provide a means for two pokemon objects to communicate with
each other through sockets. This module creates a new Comms class which inherits from
the ElectricPokemon class. Co-routines are added to the class for implenting client-server
communication through sockets and asyncio. A minor change was also made to the pokemon
class in pokemon.py.

Contributes to: DGDT-35

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+
+    Adds client-server communication between pokemon, implemented with asyncio and co-routines.
+
+    Attributes
+    ----------
+    name : str
+        Name of pokemon.
+    nickname : str
+        Nickname of pokemon.
+    moves : list
+        A list of 4 pokemon moves.
+    pokemon_type : str
+        Type of pokemon.
+    ip_address : str
+        The IP address as the client or server (default 127.0.0.1).
+    port_no : int
+        The port number to communicate on (default 8765).
+    """
```
Many of these things should be document in the base class. It's better just to document what this version adds (or removes). Otherwise you'll end up having to maintain two (or more!) sets of documentation.

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+    """
+
+    def __init__(self, name, nickname, pokemon_type, ip_address="127.0.0.1", port_no=8765):
+        """Initialise Comms class object.
+
+        Initialise with the arguments name, nickname and type, and with
+        optional keyward arguments ip_address and port_no.
+
+        Parameters
+        ----------
+        name : str
+            Name of pokemon.
+        nickname : str
+            Nickname of pokemon.
+        pokemon_type : str
+            Type of pokemon.
+        ip_address : str
+            The IP address as the client or server (default 127.0.0.1).
+        port_no : int
+            The port number to communicate on (default 8765).
+        """
```
I'd prefer if the `__init__()` doesn't get a docstring. This should go into the documentation of the class itself.  The current pre-commit flow doesn't check `__init__()` and other double-underscore methods.

Otherwise, again, there's a lot of repetition and you need to maintain documentation in multiple places.

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+        super(Comms, self).__init__(name, nickname, pokemon_type)
+
+    async def msg_tx(self, message, writer):
+        """Co-routine writes message to socket.
```
Imperative mood.
```suggestion
        """Write a message to a socket.
```

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+        await writer.drain()
+
+    async def list_tasks(self, writer):  # noqa: D301
+        """Co-routine create a random number of tasks for writing messages to socket.
```
```suggestion
        r"""Create a random number of tasks for writing messages to socket.
```

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+        writer :  StreamWriter object
+            Object instance of StreamWriter class for writing to socket.
+        """
+        print(f"Sent: {message!r}")
```
Nit-pick, but it's worth noting that you haven't actually sent the message yet. If something interrupts the program between this `print` statement and the actual `writer.drain()`, then you say that you've sent something but you don't know whether you actually have.

So I'd either change it to `Sending: `, or move it to after the `await`, or both, if you really want verbose debugging!

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+        writer.write(message.encode())
+        await writer.drain()
+
+    async def list_tasks(self, writer):  # noqa: D301
```
I don't think it's necessary to avoid D301. Use a `r` at the beginning of your docstring, otherwise the `\n` will disappear when you try to print it. This makes it a "raw" string, rather than one that gets parsed in any way before printing.

Also, this function's name doesn't seem to describe what it does too well. Perhaps something like `generate_random_tasks()`?

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+
+        Parameters
+        ----------
+        writer :  StreamWriter object
```
I prefer to move the type hints to the actual function definition where `mypy` and your IDE can parse it. Then you have something like
```python
async def list_tasks(self, writer: asyncio.StreamWriter):
  """ ...
  writer
    Handles writing to the socket.
  """
```
It just looks cleaner, it helps your tools do their job properly, and it's easier to maintain.

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+        await self.msg_tx("EOM\n", writer)
+
+    async def read_data(self, reader):
+        """Co-routine reads from socket.
```
```suggestion
        """Read from socket.
```

If you really want to make *more* explicit that it's a co-routine (though `async def` does this just fine already), then you can include the adverb "asynchronously" in your docstring, perhaps like this:

```suggestion
        """Read asynchronously from socket.
```

james-smith-za commented on communication/comms.py:
```diff
@@ -0,0 +1,185 @@
+            msg_list.append(data)
+            if data.decode() == "EOM\n":
+                print("EOM received and breaking out.")
+                print(f"Number of messages received = {len(msg_list) - 1}")
```
You wouldn't need to do the `-1` if you moved the `msg_list.append(data)` after the end of the `if` statement.

james-smith-za reviewed (commented):
Overall looking pretty good! 

The comments I made are mainly style ones. I think they are worth considering. You can make changes and then merge, or just leave it, depending on how much time and enthusiasm for the topic you have.

ijassiem committed:
Implement changes relating to comments in pull request.

These changes pertains to the comments indicated in the pull request: Create module for providing client-server communication between pokemon. #6

Contributes to: DGDT-35
